### DISCUSSION TITLE
Docs on DNS config

### DISCUSSION THREAD

**Discussion by Synchro** [laravel/valet#1380] [Q&A] [answered]
As far as I can see, there are no docs at all on the DNS config that valet wants, other than "it uses DNSMasq". In order for any local DNS server to work, it needs to be pointed at by the current networking config, so the installation process must make changes in order to make that happen. I can see that DNSMasq is running, but my networking config doesn't use localhost for DNS because it gets its DNS servers from DHCP, so it can't work. This is the default macOS config, so I can't imagine I'm alone in having DNS configured this way.

The net result is `.test` domains don't resolve, so valet isn't usable. I'm expecting that the DNSMasq config would insert itself in this lookup chain, so that the local network config points at localhost, then DNSMasq uses DHCP to find its upstream servers to resolve things other than `*.test`.

Where can I find docs on what Valet expects for its DNS config?

**Comment by drbyte** [comment, marked as answer]
Here's what Valet does for DNS:

1. Upon installation it configures `dnsmasq` to support placing custom `dnsmasq` configs into `~/.config/valet/dnsmasq.d` 
and then internally runs `valet tld test` to configure the `.test` TLD as default, as described below:

2. When the `valet tld [TLD]` command is run, a generic TLD config is created in the `dnsmasq` configuration:
eg, for `valet tld test` which configures `.test` as the default TLD for all valet-controlled domains:
- creates `~/.config/valet/dnsmasq.d/tld-test.conf`, containing two directives that allow dnsmasq to know how to handle `.test` lookups:
```
    address=/.test/127.0.0.1
    listen-address=127.0.0.1
```
- adds a MacOS `/etc/resolver/test` text file which specifies `nameserver 127.0.0.1` so MacOS looks to dnsmasq for first DNS resolution

---
In MacOS if you go to System Preferences->Network->Advanced->DNS and specify 127.0.0.1 as the first resolver and then add your local network's DHCP/DNS server addresses as "secondary" then it will first hit dnsmasq and speedily discover that it doesn't have any info for intranet/internet domains, and then will failover to your intranet's DNS, which then already has its own failover to external internet DNS.

Alternatively, you could simply use `dnsmasq`'s own config.d capability and add your own local DNS server there, like this:
`~/.config/valet/dnsmasq.d/intranet-dns.conf`
```
#local intranet DNS lookup
server=192.168.3.3
server=192.168.3.4
```
(where 192.168.3.3 and 3.4 are IP addresses of your local/intranet network server's DNS server. If only one address, skip the 2nd line. If you want to add even more including external lookups like cloudflare's 1.1.1.1 you can do that too)

And restart dnsmasq after any config changes via `valet restart`

**Reply by Synchro** [reply]
Excellent answer, thank you
